File: src/main/java/dev/slne/surf/essentials/commands/minecraft/WhitelistCommand.java
package dev.slne.surf.essentials.commands.minecraft;

import dev.jorel.commandapi.arguments.ArgumentSuggestions;
import dev.jorel.commandapi.exceptions.WrapperCommandSyntaxException;
import dev.jorel.commandapi.executors.NativeResultingCommandExecutor;
import dev.slne.surf.essentials.commands.EssentialsCommand;
import dev.slne.surf.essentials.utils.EssentialsUtil;
import dev.slne.surf.essentials.utils.color.Colors;
import dev.slne.surf.essentials.utils.brigadier.Exceptions;
import dev.slne.surf.essentials.utils.permission.Permissions;
import lombok.val;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.JoinConfiguration;
import net.kyori.adventure.text.event.HoverEvent;
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;
import org.bukkit.Server;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.event.player.PlayerKickEvent;

import java.util.concurrent.CompletableFuture;

public class WhitelistCommand extends EssentialsCommand {
    public WhitelistCommand() {
        super("whitelist", "whitelist <add | remove | list | reload | on | off | status> [<player>]", "Manage server whitelist");

        withPermission(Permissions.WHITELIST_PERMISSION);

        executesNative((NativeResultingCommandExecutor) (sender, args) -> query(sender.getCallee()));

        then(literal("on")
                .executesNative((NativeResultingCommandExecutor) (sender, args) -> whitelistState(sender.getCallee(), true)));

        then(literal("off")
                .executesNative((NativeResultingCommandExecutor) (sender, args) -> whitelistState(sender.getCallee(), false)));

        then(literal("reload")
                .executesNative((NativeResultingCommandExecutor) (sender, args) -> reload(sender.getCallee())));

        then(literal("user")
                .then(literal("list")
                        .executesNative((NativeResultingCommandExecutor) (sender, args) -> list(sender.getCallee())))
                .then(literal("removeAll")
                        .executesNative((NativeResultingCommandExecutor) (sender, args) -> removeAll(sender.getCallee())))
                .then(literal("add")
                        .then(offlinePlayerArgument("player")
                                .replaceSuggestions(ArgumentSuggestions.stringCollectionAsync(info -> CompletableFuture.supplyAsync(() -> {
                                    val offlinePlayers = info.sender().getServer().getWhitelistedPlayers();
                                    return Bukkit.getOnlinePlayers().stream()
                                            .filter(player -> !offlinePlayers.contains(player))
                                            .map(Player::getName)
                                            .toList();
                                })))
                                .executesNative((NativeResultingCommandExecutor) (sender, args) -> add(sender.getCallee(), args.getUnchecked("player")))))
                .then(literal("remove")
                        .then(offlinePlayerArgument("player")
                                .replaceSuggestions(ArgumentSuggestions.stringCollectionAsync(info -> CompletableFuture.supplyAsync(() ->
                                        info.sender().getServer().getWhitelistedPlayers().stream()
                                                .map(OfflinePlayer::getName)
                                                .toList())))
                                .executesNative((NativeResultingCommandExecutor) (sender, args) -> remove(sender.getCallee(), args.getUnchecked("player"))))));
    }

    private int query(CommandSender source) {
        val isEnabled = source.getServer().hasWhitelist();

        EssentialsUtil.sendSuccess(source, Component.text("Die Whitelist ist gerade ", Colors.INFO)
                .append(Component.text((isEnabled) ? "aktiviert" : "deaktiviert", Colors.VARIABLE_VALUE)));

        return 1;
    }

    private int whitelistState(CommandSender source, boolean enable) throws WrapperCommandSyntaxException {
        val server = source.getServer();

        if (server.hasWhitelist() == enable)
            throw enable ? Exceptions.ERROR_WHITELIST_ALREADY_ENABLED : Exceptions.ERROR_WHITELIST_ALREADY_DISABLED;
        server.setWhitelist(enable);
        server.setWhitelistEnforced(enable);

        EssentialsUtil.sendSuccess(source, Component.text("Die Whitelist wurde ", Colors.SUCCESS)
                .append(Component.text(enable ? "aktiviert" : "deaktiviert", Colors.TERTIARY)));
        return 1;
    }

    private int reload(CommandSender source) {
        val server = source.getServer();
        server.reloadWhitelist();
        kickUnlistedPlayers(server);

        EssentialsUtil.sendSuccess(source, "Die Whitelist wurde neu geladen");
        return 1;
    }

    public static void kickUnlistedPlayers(Server server) {
        if (server.isWhitelistEnforced() && server.hasWhitelist()) {
            val whitelistedPlayers = server.getWhitelistedPlayers();
            server.getOnlinePlayers().stream()
                    .filter(player -> !whitelistedPlayers.contains(player) && !player.isOp())
                    .forEach(player -> player.kick(LegacyComponentSerializer.legacySection().deserialize(EssentialsUtil.whitelistMessage()), PlayerKickEvent.Cause.WHITELIST));
        }
    }

    private int list(CommandSender source) {
        EssentialsUtil.sendSuccess(source, Component.text("Spieler auf der Whitelist: ", Colors.INFO)
                .append(Component.join(JoinConfiguration.commas(true), source.getServer().getWhitelistedPlayers().stream()
                        .map(offlinePlayer -> EssentialsUtil.getOfflineDisplayName(offlinePlayer)
                                .hoverEvent(HoverEvent.showText(Component.text(offlinePlayer.getUniqueId().toString(), Colors.VARIABLE_VALUE))))
                        .toList())));
        return 1;
    }

    private int add(CommandSender source, OfflinePlayer profile) throws WrapperCommandSyntaxException {
        if (profile.isWhitelisted()) throw Exceptions.ERROR_ALREADY_WHITELISTED;
        profile.setWhitelisted(true);

        EssentialsUtil.sendSuccess(source, EssentialsUtil.getOfflineDisplayName(profile)
                .append(Component.text(" wurde zur Whitelist hinzugefügt", Colors.SUCCESS)));
        return 1;
    }

    private int remove(CommandSender source, OfflinePlayer profile) throws WrapperCommandSyntaxException {
        if (!profile.isWhitelisted()) throw Exceptions.ERROR_NOT_WHITELISTED;
        profile.setWhitelisted(false);
        kickUnlistedPlayers(source.getServer());

        EssentialsUtil.sendSuccess(source, EssentialsUtil.getOfflineDisplayName(profile)
                .append(Component.text(" wurde von der Whitelist entfernt", Colors.SUCCESS)));

        return 1;
    }

    private int removeAll(CommandSender source) throws WrapperCommandSyntaxException {
        val server = source.getServer();
        int successes = 0;

        for (OfflinePlayer player : server.getWhitelistedPlayers()) {
            player.setWhitelisted(false);
            successes++;
        }

        kickUnlistedPlayers(server);

        if (successes == 0) throw Exceptions.ERROR_NOT_WHITELISTED;

        EssentialsUtil.sendSuccess(source, Component.text("Es wurden ", Colors.SUCCESS)
                .append(Component.text(successes, Colors.TERTIARY))
                .append(Component.text(" Spieler von der Whitelist entfernt", Colors.SUCCESS)));

        return successes;
    }
}

File: src/main/kotlin/dev/slne/surf/essentials/command/minecraft/WhitelistCommand.kt
package dev.slne.surf.essentials.command.minecraft

import com.github.shynixn.mccoroutine.folia.launch
import dev.jorel.commandapi.arguments.AsyncPlayerProfileArgument
import dev.jorel.commandapi.kotlindsl.*
import dev.slne.surf.essentials.plugin
import dev.slne.surf.essentials.util.permission.EssentialsPermissionRegistry
import dev.slne.surf.surfapi.bukkit.api.command.executors.anyExecutorSuspend
import dev.slne.surf.surfapi.bukkit.api.command.util.awaitAsyncPlayerProfile
import dev.slne.surf.surfapi.bukkit.api.command.util.idOrThrow
import dev.slne.surf.surfapi.core.api.font.toSmallCaps
import dev.slne.surf.surfapi.core.api.messages.CommonComponents
import dev.slne.surf.surfapi.core.api.messages.adventure.buildText
import dev.slne.surf.surfapi.core.api.messages.adventure.sendText
import dev.slne.surf.surfapi.core.api.messages.pagination.Pagination
import net.kyori.adventure.text.format.TextDecoration
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer

fun whitelistCommand() = commandTree("whitelist") {
    withPermission(EssentialsPermissionRegistry.WHITELIST_COMMAND)
    literalArgument("on") {
        anyExecutor { executor, _ ->
            if (Bukkit.hasWhitelist()) {
                executor.sendText {
                    appendErrorPrefix()
                    error("Die Whitelist ist bereits aktiviert.")
                }
                return@anyExecutor
            }

            Bukkit.setWhitelist(true)
            executor.sendText {
                appendSuccessPrefix()
                success("Die Whitelist wurde aktiviert.")
            }
        }
    }

    literalArgument("off") {
        anyExecutor { executor, _ ->
            if (!Bukkit.hasWhitelist()) {
                executor.sendText {
                    appendErrorPrefix()
                    error("Die Whitelist ist nicht aktiviert.")
                }
                return@anyExecutor
            }

            Bukkit.setWhitelist(false)
            executor.sendText {
                appendSuccessPrefix()
                success("Die Whitelist wurde deaktiviert.")
            }
        }
    }

    literalArgument("toggle") {
        anyExecutor { executor, _ ->
            if (Bukkit.hasWhitelist()) {
                Bukkit.setWhitelist(false)
                executor.sendText {
                    appendSuccessPrefix()
                    success("Die Whitelist wurde deaktiviert.")
                }
                return@anyExecutor
            } else {
                Bukkit.setWhitelist(true)
                executor.sendText {
                    appendSuccessPrefix()
                    success("Die Whitelist wurde aktiviert.")
                }
                return@anyExecutor
            }
        }
    }

    literalArgument("status") {
        anyExecutor { executor, _ ->
            if (Bukkit.hasWhitelist()) {
                executor.sendText {
                    appendInfoPrefix()
                    info("Die Whitelist ist aktuell ")
                    variableValue("aktiviert")
                    info(".")
                }
                return@anyExecutor
            } else {
                executor.sendText {
                    appendInfoPrefix()
                    info("Die Whitelist ist aktuell ")
                    variableValue("deaktiviert")
                    info(".")
                }
                return@anyExecutor
            }
        }
    }

    literalArgument("player") {
        literalArgument("add") {
            argument(AsyncPlayerProfileArgument("offlinePlayer")) {
                anyExecutorSuspend { executor, args ->
                    val profile = args.awaitAsyncPlayerProfile("offlinePlayer")
                    val id = profile.idOrThrow()
                    val offlinePlayer = Bukkit.getOfflinePlayer(id)

                    if (offlinePlayer.isWhitelisted) {
                        executor.sendText {
                            appendErrorPrefix()
                            error("Der Spieler ist bereits auf der Whitelist.")
                        }
                        return@anyExecutorSuspend
                    }

                    offlinePlayer.isWhitelisted = true

                    executor.sendText {
                        appendSuccessPrefix()
                        success("Der Spieler ")
                        variableValue(offlinePlayer.name ?: offlinePlayer.uniqueId.toString())
                        success(" wurde zur Whitelist hinzugefügt.")
                    }
                }
            }
        }

        literalArgument("remove") {
            argument(AsyncPlayerProfileArgument("offlinePlayer")) {
                anyExecutorSuspend { executor, args ->
                    val profile = args.awaitAsyncPlayerProfile("offlinePlayer")
                    val id = profile.idOrThrow()
                    val offlinePlayer = Bukkit.getOfflinePlayer(id)

                    if (!offlinePlayer.isWhitelisted) {
                        executor.sendText {
                            appendErrorPrefix()
                            error("Der Spieler ist nicht auf der Whitelist.")
                        }
                        return@anyExecutorSuspend
                    }

                    offlinePlayer.isWhitelisted = false

                    executor.sendText {
                        appendSuccessPrefix()
                        success("Der Spieler ")
                        variableValue(offlinePlayer.name ?: offlinePlayer.uniqueId.toString())
                        success(" wurde von der Whitelist entfernt.")
                    }
                }
            }
        }
        literalArgument("list") {
            integerArgument("page", optional = true) {
                anyExecutor { executor, args ->
                    val page: Int? by args

                    executor.sendText {
                        appendInfoPrefix()
                        info("Whitelist Informationen werden geladen...")
                    }

                    plugin.launch {
                        val whitelistedPlayers =
                            Bukkit.getWhitelistedPlayers().sortedByDescending { it.isOnline }

                        if (whitelistedPlayers.isEmpty()) {
                            executor.sendText {
                                appendErrorPrefix()
                                error("Es sind keine Spieler auf der Whitelist.")
                            }
                            return@launch
                        }


                        val pagination = Pagination<OfflinePlayer> {
                            title {
                                primary("Whitelist".toSmallCaps(), TextDecoration.BOLD)
                                spacer(" (${whitelistedPlayers.size})")
                            }
                            rowRenderer { row, _ ->
                                listOf(
                                    buildText {
                                        append(CommonComponents.EM_DASH)
                                        appendSpace()
                                        variableValue(row.name ?: row.uniqueId.toString())
                                        appendSpace()
                                        spacer("(")
                                        if (row.isOnline) {
                                            success("Online")
                                        } else {
                                            error("Offline")
                                        }
                                        spacer(")")
                                    }
                                )
                            }
                        }

                        executor.sendText {
                            append(pagination.renderComponent(whitelistedPlayers, page ?: 1))
                        }
                    }
                }
            }
        }
    }
}